waffel/waffel.py

```diff
@@ -21,112 +22,165 @@
+        # "startts":   Startzeit als unix epoch timestamp
+        # "end":       Endzeit in der gegebenen Zeitzone als HH:MM
+        # "endts":     Endzeit als unix epoch timestamp
+        # "title":     Titel des Event
+        # "subtitle":  Untertitel des Event (optional)
+        # "track":     discourse, music oder arts
+        #
+        # TODO: check, if timezone-naive is really OK
+        #       converting to timezone-aware datetimes shouldn't make a
+        #       difference.
+        return {
+            'start': start.strftime('%H:%M'),
+            'startts': self.dt_to_epoch(start),
+            'end': end.strftime('%H:%M'),
+            'endts': self.dt_to_epoch(end),
+            'title': title,
+            'subtitle': subtitle,
+            'track': self.track_map.get(track_id),
+        }
+
+    def get_events(self, track=None):
+        url = '%s?method=Event.detail&lang=en&year=%d' % (
+            self.api_url,
+            self.year
+        )
         if track:
             url = '%s&track=%s' % (url, track)
 
-        return self.__fetch_objects("events", url)
-
-    def get_locations(self, year, track=None):
-        url = '%s?method=Location.detail&lang=en&year=%d' % (self.api_url, year)
+        result = self.__fetch_objects("events", url)
+
+        ret = {}
+        for event in result:
+            start = self.parse_date(event['begin'])
+            end = self.parse_date(event['end'])
+            if not self.dt_within(start, end):
+                continue
+            location = self.location_map.get(event['location_id'], None)
+            if not location:
+                logger.warn(
+                    'location_id %s not found in location mapping' %
+                    event['location_id']
+                )
+                continue
+
+            events = []
+            if event['track'] in ('art',):
+                for appearance in event['apps']:
+                    app_start = self.parse_date(appearance['begin'])
+                    app_end = self.parse_date(appearance['end'])
+                    if not self.dt_within(app_start, app_end):
+                        continue
+                    events.append(self.make_event(
+                        app_start,
+                        app_end,
+                        appearance['name'],
+                        appearance['name_add'],  # TODO: name_add ok? type could also be of interest.  # noqa
+                        appearance['track'],
+                    ))
+            else:
+                events.append(self.make_event(
+                    start,
+                    end,
+                    event['title'],
+                    event['subtitle'],
+                    event['track'],
+                ))
+
+            if not events:
+                continue
+            ret.setdefault(location, [])
+            ret[location] += events
+
+        # TODO: needed?
+        # Sort for startts
```

Comment:
I think the lua code does it's own sorting so this not really needed.. but the lua code is highly time constrained as it also generates background animations and other stuff using OpenGL. Therefore it's good that the data is already sorted.

Reply:
ok